src/coreclr/vm/methodtablebuilder.cpp

```diff
@@ -4115,12 +4115,76 @@ VOID    MethodTableBuilder::InitializeFieldDescs(FieldDesc *pFieldDescList,
+
+                    // If the field type is a typeref instead of a typedef, it's probably in a different module, which means it can't be
+                    //  an indirect self reference and we don't need to check
+                    if (TypeFromToken(fieldTypeToken) == mdtTypeDef) {
+                        // Enumerate all the interfaces implemented by the field type
+                        HENUMInternalHolder hEnumInterfaceImpl(pInternalImport);
+                        hEnumInterfaceImpl.EnumInit(mdtInterfaceImpl, fieldTypeToken);
+
+                        ULONG cInterfaces = pInternalImport->EnumGetCount(&hEnumInterfaceImpl);
+                        if (cInterfaces != 0)
+                        {
+                            DWORD i;
+
+                            mdInterfaceImpl ii;
+                            for (i = 0; pInternalImport->EnumNext(&hEnumInterfaceImpl, &ii); i++)
+                            {
+                                mdTypeRef crInterface;
+                                pInternalImport->GetTypeOfInterfaceImpl(ii, &crInterface);
+                                if (crInterface == GetCl())
+                                    fIsSelfReferentialThroughInterface = TRUE;
+                            }
+                        }
+                    }
+
+                    if (fIsSelfReferentialThroughInterface)
+                    {
+                        pByValueClass = (MethodTable*)-1;
+                    }
+                }
+
                 // It's not self-referential so try to load it
                 if (pByValueClass == NULL)
                 {
+                    // Loading a static valuetype field on a struct could be indirectly self-referential if one of the field type's
+                    //  generic arguments is the valuetype we're currently loading.
+                    BOOL fCouldBeSelfReferentialThroughGenerics = fIsStatic && IsValueClass();
+                    if (fCouldBeSelfReferentialThroughGenerics)
+                    {
+                        SigPointer sptr(pMemberSignature, cMemberSignature);
+                        uint32_t unused;
+                        sptr.GetCallingConvInfo(&unused); // Skip calling convention, we already know it
+                        CorElementType rawElementType;
+                        sptr.PeekElemType(&rawElementType);
+                        if (rawElementType != ELEMENT_TYPE_GENERICINST) // Determine whether this valuetype is a generic instance
+                            fCouldBeSelfReferentialThroughGenerics = FALSE;
+                    }
+
                     // Loading a non-self-ref valuetype field.
                     OVERRIDE_TYPE_LOAD_LEVEL_LIMIT(CLASS_LOAD_APPROXPARENTS);
-                    if (isEnCField || fIsStatic)
+                    if ((isEnCField || fIsStatic) && !fCouldBeSelfReferentialThroughGenerics)
```

Comment:
Have you explored whether or not we could just remove the fIsStatic flag check here? I have some *very* vague memories that this might not be necessary from previous work in the area, but the bug I was fixing was risky enough as it was so I restricted it to the instance scenarios I was working on.

Reply:
I didn't test that, I'll check.